pkgs/development/python-modules/openai-triton/default.nix

```diff
@@ -0,0 +1,189 @@
+{ lib
+, buildPythonPackage
+, stdenv
+, fetchurl
+, fetchFromGitHub
+, cmake
+, cudaPackages ? { }
+, llvmPackages
+, pybind11
+, gtest
+, zlib
+, ncurses
+, libxml2
+, lit
+, filelock
+, torchWithRocm
+, pytest
+}:
+
+let
+  pname = "triton";
+  version = "2.0.0";
+
+  ptxas = "${cudaPackages.cuda_nvcc}/bin/ptxas";
+in
+buildPythonPackage {
+  inherit pname version;
+
+  format = "setuptools";
+
+  src = fetchFromGitHub {
+    owner = "openai";
+    repo = "triton";
+    rev = "v${version}";
+    hash = "sha256-9GZzugab+Pdt74Dj6zjlEzjj4BcJ69rzMJmqcVMxsKU=";
+  };
+
+  patches = [
+    # Prerequisite for llvm15 patch
+    (fetchurl {
+      url = "https://github.com/openai/triton/commit/2aba985daaa70234823ea8f1161da938477d3e02.patch";
+      hash = "sha256-HEuLZFif++a/fKs3dyIhqSc+D2DPbzEXOSSR4nRWtgQ=";
+    })
+    (fetchurl {
+      url = "https://github.com/openai/triton/commit/e3941f9d09cdd31529ba4a41018cfc0096aafea6.patch";
+      hash = "sha256-sl8woykSLCq4ZJYzEQdCPWM8rzv+s4RTK7PuqsTcmy8=";
+    })
+
+    # Not using fetchurl because needed to remove binary diff for ptxas
+    ./llvm15.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace python/setup.py \
+      --replace \
+        '= get_thirdparty_packages(triton_cache_path)' \
+        '= os.environ["cmakeFlags"].split()'
+    sed -i '/install_requires=/,/],/d' python/setup.py
+  ''
+  # A typo in setup.py?
+  + ''
+    substituteInPlace python/setup.py \
+      --replace '"tests"' '"test"'
+  ''
+  # Circular dependency, cf. https://github.com/openai/triton/issues/1374
+  + ''
+    substituteInPlace python/setup.py --replace '"torch",' ""
```

Comment:
The Python package docs recommend removing circular dependencies (https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#how-to-solve-circular-dependencies-how-to-solve-circular-dependencies), but they don't say how... just today in the `torch-bin` 2.0 PR I recommended this be done through the `pythonRemoveDeps` hook (https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonrelaxdepshook-using-pythonrelaxdepshook): https://github.com/NixOS/nixpkgs/pull/221652#issuecomment-1478594840

What are your thoughts on doing that vs. the `substituteInPlace` approach?

Reply:
Thanks, I didn't know `pythonRemoveDeps` exists! The documentation gives an example with `requirements.txt`, does it also work with `setup.py`?

Reply:
Using `pythonRemoveDeps` now